Data sensitivity feature, fixes # 8

Sidrah-Madiha opened the pull request:
This branch that I have merged fixes #8 ,
-  it  contains a helper function file "helper_for_senstivity_calculation.py" which has a function "calculate_senstivity" that returns L1 sensitivity of the dataset and list of performance metric for all parallel dataset
-  I have also added a function to display the most sensitive datapoint as well as the index of that data point in training dataset.
- I have tested it on the random forest classifier that I made for vehicles.csv dataset, please see test demo in jupyer notebook in folder: " Importance score for dataset training samples" >  vehicles_dataset_classifer_v1.ipynb file

@dzeber  I was not sure if the performance metric should be accuracy or score so I implemented it with score and commented out code that I used with accuracy . Please advise for the next steps to improve this fix.

dzeber commented on dev/Sidrah-Madiha/Importance score for dataset training samples/helper_for_senstivity_calculation.py:
```diff
@@ -0,0 +1,57 @@
+    model_func.fit(X_train, y_train)
+    # y_pred=model_func.predict(X_test) 
+    # return metrics.accuracy_score(y_test, y_pred)
+    # I am not sure if accuracy or score is record for senstivity calculation so I commented out accuracy code
```
I'm pretty sure these do the same think. I thing `model_func.score()` is just a shortcut.

dzeber commented on dev/Sidrah-Madiha/Importance score for dataset training samples/vehicles_dataset_classifer_v1.ipynb:
```diff
@@ -0,0 +1,826 @@
+    }
+   ],
+   "source": [
+    "print(\"priniting model performance score for all parallel datasets\", list_scores)\n",
```
Could you show a visualization instead of the full list?

dzeber commented on dev/Sidrah-Madiha/Importance score for dataset training samples/helper_for_senstivity_calculation.py:
```diff
@@ -0,0 +1,57 @@
+    list_train_data_label =create_all_parallel_dbs(X_train, y_train) #return list containing tuples of training data and label
+    #print(len(list_train_data_label ))
+    for parallel_train, parallel_label in list_train_data_label:
+        parallel_dataset_score =model_performance_evaluater_score(model_func,parallel_train, parallel_label, X_test, y_test)
```
Rather than precomputing all the parallel DBs, why not call `create_parallel_db()` inside this loop and compute the score right away for each. This will probably be easier for larger datasets.

dzeber commented on dev/Sidrah-Madiha/Importance score for dataset training samples/helper_for_senstivity_calculation.py:
```diff
@@ -0,0 +1,57 @@
+    #print(len(list_train_data_label ))
+    for parallel_train, parallel_label in list_train_data_label:
+        parallel_dataset_score =model_performance_evaluater_score(model_func,parallel_train, parallel_label, X_test, y_test)
+        dataset_distance = np.abs(full_dataset_score - parallel_dataset_score ) # L1 senstivity
```
I'm thinking it would be interesting compute the distance without the absolute value. It would probably be valuable to know if certain training samples actually improve the model fit when being left out.

mlopatka commented:
@Sidrah-Madiha please let us know if you intend to continue with this PR or if this work has been integrated to another PR and this one can be closed out.

Sidrah-Madiha commented:
@mlopatka , yes I am working on fixing it today.

Sidrah-Madiha commented:
@mlopatka please accept my PR for this issue, I have fied all change requests.
- reformatted code for helper file with black.
- showed visualization of full list of scores for each parallel db. Please see image attached
![sens2](https://user-images.githubusercontent.com/16177747/76843605-de5b6400-685d-11ea-91ca-f23dbac24b4e.png)
- fixed this issue _(Rather than precomputing all the parallel DBs, why not call create_parallel_db() inside this loop and compute the score right away for each. This will probably be easier for larger datasets.)_ please see attached image for changes for
![sens1](https://user-images.githubusercontent.com/16177747/76843737-06e35e00-685e-11ea-8ca3-982bc788b962.png)
 code refactoring
- removed absolute function from senstivity distance difference calculation. (showed in above iage aas well)

Sidrah-Madiha commented:
Hi @dzeber, please merge this PR, I have fixed all change requests

Sidrah-Madiha commented:
Hi @dzeber I have further improved this fix, I have added a scoring parameter with default "accuracy_score in function "model_performance_evaluater_score"
I have fixed relative path issue in notebook
I have added an indicator in the visualization to indicate the most sensitive datapoint, fixed labels to have better description. 
![sensitive](https://user-images.githubusercontent.com/16177747/77259657-dc9e0000-6ca4-11ea-89de-fbb7ca9cc8ab.png)

Please review
